distributed/scheduler.py

```diff
@@ -2315,19 +2327,49 @@ def decide_worker(self, ts: TaskState) -> WorkerState:
         """
         Decide on a worker for task *ts*.  Return a WorkerState.
         """
+        if not self._workers_dv:
+            return None
+
         ws: WorkerState = None
+        group: TaskGroup = ts._group
         valid_workers: set = self.valid_workers(ts)
 
         if (
             valid_workers is not None
             and not valid_workers
             and not ts._loose_restrictions
-            and self._workers_dv
         ):
             self._unrunnable.add(ts)
             ts.state = "no-worker"
             return ws
 
+        # Group is larger than cluster with few dependencies? Minimize future data transfers.
+        if (
+            valid_workers is None
+            and len(group) > self._total_nthreads * 2
+            and sum(map(len, group._dependencies)) < 5
```

Comment:
I'm wondering if it is possible to construct a test which helps us with this cutoff condition. One of the motivators for this heuristics is that this is likely acting as a good binary classifier for graphs where most have either very small or very large numbers here. afaiu, this is an unproven assumption

I'm interested in a test for this particular boundary condition for two reasons
* It would helps us to identify regressions when this boundary is accidentally moved
* How good is our classifier? Might it be 40% better if this value was 10? A well written test could help in an analysis for this

I won't push hard on this if it proves too difficult or others disagree on the value. I'm just having a hard time with heuristics if I can change them without tests breaking.

For the current test I can do either of the following without the test breaking

* remove `len(group) > self._total_nthreads * 2` entirely
* Increase the boundary for total thread count, e.g. `len(group) > self._total_nthreads * 10` (it breaks eventually if pushed further)
* Increase the boundary for dependencies `sum(map(len, group._dependencies)) < 100` (increased further, it breaks.

Reply:
In 90% of cases the number here is 0 (like `da.random.random`) or 1 (like `da.from_zarr`).  I can *imagine* but can't actually come up with cases where this might be 2 (like `da.from_zarr(zarr_array, parameter=some_dask_thing)`)

I get the aversion to magic numbers.  This one feels pretty safe to me though.